docs/source/notes/randomness.rst

```diff
@@ -2,75 +2,115 @@ Reproducibility
+
+CUDA convolution benchmarking
+-----------------------------
+CUDA convolution operations have a performance enhancement feature that can introduce
+randomness between multiple executions of an application. You can disable this feature
+with :code:`torch.backends.cudnn.benchmark = False`.
+
+However, if you do not need reproducibility across multiple executions of your application,
+then it would be beneficial to enable this benchmarking feature.
+
+Any time a CUDA convolution operation is called with a new set of size parameters and
+:code:`torch.backends.cudnn.benchmark = True` is set, CUDA will run multiple different
+implementations of the operation and benchmark them. The fastest implementation for your
+patform will then be cached and used for all further CUDA convolution operations that
+use that same set of size parameters, until the PyTorch program quits. Since the runtime
+of a convolution can vary, the fastest implementation could change each time the
+benchmark is run.
+
+Note that this setting is different from the :code:`torch.backends.cudnn.deterministic`
+setting discussed below.
+
+Avoiding nondeterministic algorithms
+....................................
+:meth:`torch.set_deterministic` lets you configure PyTorch to use deterministic
+algorithms instead of nondeterministic ones where available, and to throw an error
+if an operation is known to be nondeterministic (and without a deterministic
+alternative).
+
+Please check the documentation for :meth:`torch.set_deterministic()` for a full
+list of affected operations. If an operation does not act correctly according to
+the documentation, or if you need a deterministic implementation of an operation
+that does not have one, please submit an issue:
+`<https://github.com/pytorch/pytorch/issues?q=label:%22topic:%20determinism%22>`_
+
+For example, running the nondeterministic CUDA implementation of :meth:`index_add_`
+will throw an error::
+
+    >>> import torch
+    >>> torch.set_deterministic(True)
+    >>> torch.randn(2, 2).cuda().index_add_(0, torch.tensor([0, 1]), torch.randn(2, 2))
+    Traceback (most recent call last):
+    File "<stdin>", line 1, in <module>
+    RuntimeError: index_add_cuda_ does not have a deterministic implementation, but you set
+    'torch.set_deterministic(True)'. ...
+
+Furthermore, if you are using CUDA tensors, and your CUDA version is 10.2 or greater, you
+should set the environment variable `CUBLAS_WORKSPACE_CONFIG=:4096:8` or
+`CUBLAS_WORKSPACE_CONFIG=:16:8`. See :meth:`torch.set_deterministic` documentation
+for more details.
+
+CUDA convolution determinism
+----------------------------
+When running convolution operations on CUDA devices, a nondeterministic algorithm
+could potentially be used. To avoid this, you can set
+:code:`torch.backends.cudnn.deterministic = True`. However, setting
+:code:`torch.set_deterministic(True)` will also prevent CUDA convolution from using
```

Comment:
I would promote this option first. Following the example lead in it could go something like,

"... nondeterministic unless `torch.set_deterministic(True)` or `torch.backends.cudnn.deterministic = True`." The latter setting controls only this behavior, unlike `torch.set_deterministic(True)` which will make other PyTorch operations behave deterministically, too.